src/lib.rs

```diff
@@ -187,20 +194,36 @@ impl<D> EventSource for WaylandSource<D> {
             }
         }
 
-        // ensure we are not waiting for messages that are already in the buffer.
-        Self::loop_callback_pending(&mut self.queue, &mut callback)?;
-        self.read_guard = Some(Self::prepare_read(&mut self.queue)?);
-
-        Ok(())
+        self.read_guard = self.queue.prepare_read();
+        match self.read_guard {
+            Some(_) => Ok(None),
+            // If getting the guard failed, that means that there are
+            // events in the queue, and
+            // The readiness value is never used, we just need some marker
+            // If getting the guard failed, we need to process the events 'instantly'
+            // tell calloop this
+            None => Ok(Some((Readiness::EMPTY, self.fake_token.unwrap()))),
+        }
     }
 
-    fn post_run<F>(&mut self, _: F) -> calloop::Result<()>
-    where
-        F: FnMut((), &mut Self::Metadata) -> Self::Ret,
-    {
-        // Drop implementation of ReadEventsGuard will do cleanup
-        self.read_guard.take();
-        Ok(())
+    fn before_handle_events(&mut self, events: calloop::EventIterator<'_>) {
+        let guard = self.read_guard.take();
+        if events.count() > 0 {
+            // 1. read events from the socket if any are available
+            if let Some(guard) = guard {
+                if let Err(WaylandError::Io(err)) = guard.read() {
```

Comment:
You can probably merge these two and probably, but I don't really care.

Reply:
I don't really see how without let chains, which I don't think are stable yet

Reply:
I've suggested in the diff how to merge them.

Reply:
I see. It's not the code style I would use, but I'll apply it anyway.